File: NET/JNetCall.Sharp/Server/AbstractHost.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using JNetCall.Sharp.API;
using JNetProto.Sharp.Beans;
using JNetProto.Sharp.Tools;

namespace JNetCall.Sharp.Server
{
    public abstract class AbstractHost<T> : IDisposable
    {
        private readonly Type _serviceClass;
        private readonly IDictionary<string, Type> _interfaces;
        protected readonly ProtoSettings _config;

        protected AbstractHost(Type serviceClass)
        {
            _serviceClass = serviceClass;
            _interfaces = new Dictionary<string, Type>();
            _config = new ProtoSettings();
        }

        protected T CreateInst()
        {
            var svc = (T)Activator.CreateInstance(_serviceClass);
            return svc;
        }

        public void AddServiceEndpoint<TSvc>()
        {
            AddServiceEndpoint(typeof(TSvc));
        }

        private void AddServiceEndpoint(Type interfaceClass)
        {
            var name = interfaceClass.Name;
            _interfaces.Add(name, interfaceClass);
        }

        protected void HandleCall(object inst, MethodInfo[] methods, 
            MethodCall call, ProtoConvert proto)
        {
            if (!_interfaces.ContainsKey(call.C))
            {
                var debug = call.C;
                Write(proto, debug, MethodStatus.ClassNotFound);
                return;
            }
            var callName = call.M;
            var method = methods.FirstOrDefault(m => CheckMethod(m, callName));
            if (method == null)
            {
                var debug = call.C + "::" + call.M;
                Write(proto, debug, MethodStatus.MethodNotFound);
                return;
            }
            try
            {
                var types = method.GetParameters()
                    .Select(p => p.ParameterType).ToArray();
                var args = Conversions.Convert(types, call.A);
                var res = method.Invoke(inst, args);
                Write(proto, res, MethodStatus.Ok);
            }
            catch (Exception e)
            {
                var cause = e is TargetInvocationException ti
                    ? ti.InnerException
                    : e;
                var debug = cause!.ToString();
                Write(proto, debug, MethodStatus.MethodFailed);
            }
        }

        private const StringComparison Cmp = StringComparison.InvariantCultureIgnoreCase;

        private static bool CheckMethod(MemberInfo m, string callName)
        {
            var mName = m.Name.Replace("_", string.Empty);
            return mName.Equals(callName, Cmp);
        }

        private static void Write(ProtoConvert proto, object res, MethodStatus status)
        {
            var obj = new MethodResult(res, (short)status);
            proto.WriteObject(obj);
            proto.Flush();
        }

        public void Dispose()
        {
            _interfaces.Clear();
        }
    }
}

File: JVM/jnetcall-java/src/main/java/jnetcall/java/server/AbstractHost.java
package jnetcall.java.server;

import jnetcall.java.api.MethodCall;
import jnetcall.java.api.MethodResult;
import jnetcall.java.api.MethodStatus;
import jnetproto.java.beans.ProtoConvert;
import jnetproto.java.beans.ProtoSettings;
import jnetproto.java.compat.Strings;
import jnetproto.java.tools.Conversions;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

abstract class AbstractHost<T> implements AutoCloseable {
    private final Class<T> serviceClass;
    private final Map<String, Class<?>> interfaces;
    protected final ProtoSettings config;

    protected AbstractHost(Class<T> serviceClass) {
        this.serviceClass = serviceClass;
        this.interfaces = new HashMap<>();
        this.config = new ProtoSettings();
    }

    protected T createInst() throws Exception {
        var clazz = this.serviceClass;
        var svc = clazz.getDeclaredConstructor().newInstance();
        return svc;
    }

    public <I> void addServiceEndpoint(Class<I> interfaceClass) {
        var name = interfaceClass.getSimpleName();
        this.interfaces.put(name, interfaceClass);
    }

    protected void handleCall(Object inst, Method[] methods, MethodCall call, ProtoConvert proto)
            throws Exception {
        if (!interfaces.containsKey(call.C())) {
            var debug = call.C();
            write(proto, debug, MethodStatus.ClassNotFound);
            return;
        }
        var callName = call.M();
        var method = Arrays.stream(methods)
                .filter(m -> checkMethod(m, callName))
                .findFirst().orElse(null);
        if (method == null) {
            var debug = call.C() + "::" + call.M();
            write(proto, debug, MethodStatus.MethodNotFound);
            return;
        }
        try {
            var args = Conversions.convertFor(call.A(), method);
            var res = method.invoke(inst, args);
            write(proto, res, MethodStatus.Ok);
        } catch (Throwable e) {
            var cause = e instanceof InvocationTargetException
                    ? e.getCause() : e;
            var debug = Strings.getStackTrace(cause);
            write(proto, debug, MethodStatus.MethodFailed);
        }
    }

    private boolean checkMethod(Method m, String callName) {
        var cName = callName.replace("_", "");
        return m.getName().equalsIgnoreCase(cName);
    }

    private static void write(ProtoConvert proto, Object res, MethodStatus status)
            throws Exception {
        var obj = new MethodResult(res, status.getValue());
        proto.writeObject(obj);
        proto.flush();
    }

    @Override
    public void close() throws Exception {
        this.interfaces.clear();
    }
}